agent/configuration-build-derive/src/derive.rs

```diff
@@ -0,0 +1,434 @@
+use proc_macro2::{Ident, Span, TokenStream};
+use syn::{self, Data, LitStr};
+
+extern crate quote;
+
+pub(crate) fn impl_configuration(ast: &syn::DeriveInput) -> TokenStream {
+    let builder_name = format!("{}{}", &ast.ident.to_string(), "Builder");
+    let builder_ident = Ident::new(&builder_name, Span::call_site());
+    let ident = &ast.ident;
+    quote! {
+       impl model::Configuration for #ident{}
+
+       impl #ident{
+           pub fn builder() -> #builder_ident {
+               #builder_ident::new()
+           }
+       }
+    }
+}
+
+pub(crate) fn build_struct(ast: &syn::DeriveInput) -> TokenStream {
+    let name = format!("{}{}", &ast.ident.to_string(), "Builder");
+    let build_ident = Ident::new(&name, Span::call_site());
+    let data = match &ast.data {
+        Data::Struct(data) => data,
+        _ => panic!("configuration-derive only supports structs"),
+    };
+    let crd_type = ast
+        .attrs
+        .iter()
+        .filter_map(|v| {
+            v.parse_meta().ok().map(|meta| {
+                if meta.path().is_ident("crd") {
+                    v.parse_args::<LitStr>().ok()
+                } else {
+                    None
+                }
+            })
+        })
+        .last()
+        .flatten()
+        .expect("`crd` is a required attribute (Test, Resource)")
+        .value();
+
+    // Get a list of fields and their types
+    let fields = data.fields.iter().filter_map(|field| {
+        let doc = field.attrs.iter().filter(|v| {
+            v.parse_meta()
+                .map(|meta| meta.path().is_ident("doc"))
+                .unwrap_or(false)
+        });
+        let field_name = match field.ident.as_ref() {
+            Some(ident) => ident.to_string(),
+            None => return None,
+        };
+        let field_ident = Ident::new(&field_name, Span::call_site());
+        let ty = field.ty.clone();
+        Some(quote! {
+            #(#doc)*
+            #field_ident: model::ConfigValue<#ty>,
+        })
+    });
+    // Create the setters for each field, one for typed values and one for templated strings
+    let setters = data.fields.iter().filter_map(|field| {
+        let doc = field.attrs.iter().filter(|v| {
+            v.parse_meta()
+                .map(|meta| meta.path().is_ident("doc"))
+                .unwrap_or(false)
+        });
+        let field_name = match field.ident.as_ref() {
+            Some(ident) => ident.to_string(),
+            None => return None,
+        };
+        let field_ident = Ident::new(&field_name, Span::call_site());
+        let template_field_name = format!("{}{}", field_name, "_template");
+        let template_ident = Ident::new(&template_field_name, Span::call_site());
+        let ty = field.ty.clone();
```

Comment:
I don't like it, but it's definitely not worth blocking over. Just because somebody did a bad thing and used a variable named `ty` doesn't mean I should. If everyone jumped off a cliff...